src/sentry/search/snuba/backend.py

```diff
@@ -231,76 +220,119 @@ def _query(self, project, retention_window_start, group_queryset, tags, environm
-            # If the pre-filter query didn't include anything to significantly
-            # filter down the number of results (from 'first_release', 'query',
-            # 'status', 'bookmarked_by', 'assigned_to', 'unassigned',
-            # 'subscribed_by', 'active_at_from', or 'active_at_to') then it
-            # might have surpassed the MAX_PRE_SNUBA_CANDIDATES. In this case,
-            # we *don't* want to pass candidates down to Snuba, and instead we
-            # want Snuba to do all the filtering/sorting it can and *then* apply
-            # this queryset to the results from Snuba, which we call
-            # post-filtering.
-            metrics.incr('snuba.search.too_many_candidates')
-            candidate_hashes = None
+        candidate_hashes = None
+        if max_pre_snuba_candidates:
+            candidate_hashes = dict(
+                GroupHash.objects.filter(
+                    group__in=group_queryset
+                ).values_list(
+                    'hash', 'group_id'
+                )[:max_pre_snuba_candidates + 1]
+            )
+            metrics.timing('snuba.search.num_candidates', len(candidate_hashes))
+
+            if not candidate_hashes:
+                # no matches could possibly be found from this point on
+                metrics.incr('snuba.search.no_candidates')
+                return Paginator(Group.objects.none()).get_result()
+            elif len(candidate_hashes) > max_pre_snuba_candidates:
+                # If the pre-filter query didn't include anything to significantly
+                # filter down the number of results (from 'first_release', 'query',
+                # 'status', 'bookmarked_by', 'assigned_to', 'unassigned',
+                # 'subscribed_by', 'active_at_from', or 'active_at_to') then it
+                # might have surpassed the `max_pre_snuba_candidates`. In this case,
+                # we *don't* want to pass candidates down to Snuba, and instead we
+                # want Snuba to do all the filtering/sorting it can and *then* apply
+                # this queryset to the results from Snuba, which we call
+                # post-filtering.
+                metrics.incr('snuba.search.too_many_candidates')
+                candidate_hashes = None
 
         sort, extra_aggregations, score_fn = sort_strategies[sort_by]
+        chunk_limit = limit
+        offset = 0
+        num_chunks = 0
+        paginator_results = Paginator(Group.objects.none()).get_result()
+        result_groups = []
+
+        # Do smaller searches in chunks until we have enough results
+        # to answer the query (or hit the end of possible results). We do
+        # this because a common case for search is to return 100 groups
+        # sorted by `last_seen`, and we want to avoid returning all of
+        # a project's hashes and then post-sorting them all in Postgres
+        # when typically the first N results will do.
+        while True:
```

Comment:
There are apparently projects with millions of groups. I wonder if this should be `while num_chunks < 100` (or something) and as a separate/future change we could bubble up some kind of better response to tell them they need to hone in on their search?

Reply:
Is this just the case where lots of hashes come back, but very few of them match the groups we want after filtering? so we're stuck returning chunk after chunk from snuba forever and discarding them?

Reply:
Right, imagine a project with a million groups and they want to find the needle in the haystack. Or I guess, `>= MAX_CANDIDATES + 1` in the haystack since that would handle it otherwise.